fix: insert distributed table if partition column has default value

NiwakaDev opened the pull request:
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

_PLEASE DO NOT LEAVE THIS EMPTY !!!_

This PR allows `greptime` to insert data into a distributed table without explicitly setting the partition column value if it has a default value.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)

#1293

WenyXu commented on src/partition/src/splitter.rs:
```diff
@@ -42,12 +43,35 @@ impl WriteSplitter {
+        let mut insert = insert;
+        let partition_columns = self.partition_rule.partition_columns();
+        for column_schema in schema.column_schemas() {
+            let key = insert.columns_values.keys().next().unwrap();
+            let row_nums = insert.columns_values.get(key).unwrap().len();
```
Is this will panic at where `schema.column_schemas().len()` != `insert.columns_values.keys().len()`?

NiwakaDev replied:
I guess that this doesn't panic because the following test passes:

https://github.com/GreptimeTeam/greptimedb/pull/1498/files#diff-34bcab2e99e5f158ebeceb2c917d7495ea0538ec0312ee1af1aaaf4c337dd887R302-R305

In the above test, `schema.column_schemas().len()` is not equal to `insert.columns_values.keys().len()`.

WenyXu commented on src/partition/src/splitter.rs:
```diff
@@ -277,7 +302,8 @@ mod tests {
         let rule = Arc::new(MockPartitionRule) as PartitionRuleRef;
         let spliter = WriteSplitter::with_partition_rule(rule);
-        let ret = spliter.split_insert(insert).unwrap();
+        let stub_schema = DataTypesSchema::new(vec![]);
+        let ret = spliter.split_insert(insert, &stub_schema).unwrap();
```
How about adding more unit tests?

WenyXu commented on tests/cases/standalone/common/basic.result:
```diff
@@ -58,3 +58,25 @@ DROP TABLE system_metrics;
 
 Affected Rows: 1
 
+create table foo (
+    host string,
+    ts timestamp DEFAULT CURRENT_TIMESTAMP,
+    cpu double default 0,
+    TIME INDEX (ts),
+    PRIMARY KEY(host)
+) engine=mito with(regions=1);
+
+Affected Rows: 0
+
+insert into foo (host, cpu, ts) values ('host1', 1.1, 1000);
+
+Affected Rows: 1
+
+insert into foo (host, cpu) values ('host2', 2.2);
+
+Affected Rows: 1
+
+DROP TABLE foo;
+
+Affected Rows: 1
```
It's better to use `select * from foo` to verify inserted default value.

WenyXu commented:
Hi NiwakaDev, thank you for your contribution. My colleagues are on vacation during Golden Week and may be slow to respond.

NiwakaDev commented:
@WenyXu 

Thank you for your quick review!

Addressed your review. Please let me know if there are any errors or omissions.

>My colleagues are on vacation during Golden Week and may be slow to respond.

Ok.

MichaelScofield commented on src/partition/src/splitter.rs:
```diff
@@ -42,12 +43,35 @@ impl WriteSplitter {
+            if partition_columns.contains(&column_schema.name)
+                && (!insert.columns_values.contains_key(&column_schema.name))
+            {
+                let key = insert.columns_values.keys().next().unwrap();
+                let row_nums = insert.columns_values.get(key).unwrap().len();
```
The row numbers are checked in method `check_req` above. Can you return the row number from `check_req` and use it here?

MichaelScofield commented on src/partition/src/error.rs:
```diff
@@ -136,6 +150,9 @@ impl ErrorExt for Error {
             Error::InvalidTableRouteData { .. } => StatusCode::Internal,
             Error::ConvertScalarValue { .. } => StatusCode::Internal,
             Error::FindDatanode { .. } => StatusCode::InvalidArguments,
+            Error::CreateDefaultToRead { .. } | Error::MissingDefaultValue { .. } => {
```
I think we can use `CreateDefaultToRead`'s `source`'s status, like this: `Error::CreateDefaultToRead { source, .. } => source.status_code()`

WenyXu commented on tests/cases/standalone/common/basic.sql:
```diff
@@ -24,3 +24,19 @@ SELECT avg(cpu_util) FROM system_metrics;
+insert into foo (host, cpu, ts) values ('host1', 1.1, '2000-01-01 00:00:00+00:00');
+
+insert into foo (host, cpu) values ('host2', 2.2);
+
```
```suggestion

insert into foo (host) values ('host2');
```

NiwakaDev commented on src/partition/src/splitter.rs:
```diff
@@ -42,12 +43,35 @@ impl WriteSplitter {
-        let partition_columns = find_partitioning_values(values, &column_names)?;
+        let mut insert = insert;
+        let partition_columns = self.partition_rule.partition_columns();
+        for column_schema in schema.column_schemas() {
+            if partition_columns.contains(&column_schema.name)
+                && (!insert.columns_values.contains_key(&column_schema.name))
+            {
+                let key = insert.columns_values.keys().next().unwrap();
+                let row_nums = insert.columns_values.get(key).unwrap().len();
+                let default_values = column_schema
+                    .create_default_vector(row_nums)
+                    .context(CreateDefaultToReadSnafu {
+                        column: &column_schema.name,
+                    })?
+                    .context(MissingDefaultValueSnafu {
+                        column: &column_schema.name,
+                    })?;
+                insert
+                    .columns_values
+                    .insert(column_schema.name.clone(), default_values);
+            }
```
I get the following error:

```
cannot borrow `insert.columns_values` as mutable because it is also borrowed as immutable
mutable borrow occurs hererustc ~
//...
splitter.rs(57, 21): immutable borrow occurs here
splitter.rs(57, 73): first borrow occurs due to use of `insert.columns_values` in closure
splitter.rs(59, 30): immutable borrow later used here
```

So, how about this instead?

```rust
        let partition_column_schemas = schema
            .column_schemas()
            .iter()
            .filter(|schema| partition_columns.contains(&schema.name));
        for partition_column_schema in partition_column_schemas {
            if !insert.columns_values.contains_key(&partition_column_schema.name){
                let default_values = partition_column_schema
                    .create_default_vector(row_nums)
                    .context(CreateDefaultToReadSnafu {
                        column: &partition_column_schema.name,
                    })?
                    .context(MissingDefaultValueSnafu {
                        column: &partition_column_schema.name,
                    })?;
                insert
                    .columns_values
                    .insert(partition_column_schema.name.clone(), default_values);
            }
        }
```

WenyXu replied:
> I get the following error:

That was my mistake. Try this. It's worked for me
```rust
            .filter(|schema| partition_columns.contains(&schema.name) && !insert.columns_values.contains_key(&schema.name))
            .collect::<Vec<_>>();
```

WenyXu commented:
lgtm

WenyXu commented on src/partition/src/splitter.rs:
```diff
@@ -302,8 +302,79 @@ mod tests {
+            r1_columns.get("id").unwrap().get(0)
+        );
+        assert_eq!(
+            <&str as Into<Value>>::into("host1"),
```
It's preferable to use `Value::from`

WenyXu commented on src/partition/src/splitter.rs:
```diff
@@ -277,7 +308,16 @@ mod tests {
+            ColumnSchema::new("id", ConcreteDataType::Int16(Int16Type {}), false),
+            ColumnSchema::new("host", ConcreteDataType::String(StringType), true),
+        ]);
+        let ret = spliter.split_insert(insert, &mock_schema).unwrap();
```
I realized your test; actually won't run into your code. In the test self.partition_rule.partition_columns() returns ["id"]. However, the `insert` id columns are non-null values.

WenyXu commented on src/partition/src/splitter.rs:
```diff
@@ -327,6 +367,54 @@ mod tests {
     }
 
+    #[test]
+    fn test_writer_spliter_when_insert_columns_len_is_not_equal_to_schema_columns_len() {
```
We can safely remove this test because we already removed the code potential to cause panic.

WenyXu reviewed (approved):
Hi Niwaka, It looks good to me. Thank you for your contribution again 🥳

nit: I introduced a helper function to assert the values in columns.

MichaelScofield commented:
thx @NiwakaDev

NiwakaDev commented:
@WenyXu , @MichaelScofield 
Thank you very much for your review! I'll do my best to contribute more to greptime!
